### PR TITLE
fix moving files from external storage to object store trashbin

### PR DESCRIPTION
**Pull request by icewind1991**
having the "cache rename" after the "storage move" caused the target
to get the fileid from the source file, without taking care that the object
is stored under the original file id.

By doing the "cache rename" first, we trigger the "update existing file"
logic while moving the file to the object store and the object gets stored for the
correct file id

Signed-off-by: Robin Appelman <robin@icewind.nl>

### CODE REVIEW

**Comment by kesselb** [comment]
cc @Dragon988

**Comment by Dragon988** [comment]
I'm not using S3 anymore but the code looks good to me.

File: lib/private/Files/Storage/Common.php [hunk 1]
```diff
@@ -620,18 +620,15 @@ public function copyFromStorage(IStorage $sourceStorage, $sourceInternalPath, $t
...
 			$source = $sourceStorage->fopen($sourceInternalPath, 'r');
-			// TODO: call fopen in a way that we execute again all storage wrappers
-			// to avoid that we bypass storage wrappers which perform important actions
-			// for this operation. Same is true for all other operations which
-			// are not the same as the original one.Once this is fixed we also
-			// need to adjust the encryption wrapper.
-			$target = $this->fopen($targetInternalPath, 'w');
-			[, $result] = \OC_Helper::streamCopy($source, $target);
+			if ($source) {
+				$result = $this->writeStream($targetInternalPath, $source) > 0;
```
Comment (rullzer) [inline lib/private/Files/Storage/Common.php:624]: would this not do boom if you work with empty files?

Comment (icewind1991) [reply lib/private/Files/Storage/Common.php]: switched to exceptions to remove this ambiguity

File: lib/private/Files/Storage/Local.php [hunk 2]
```diff
@@ -553,6 +554,11 @@ public function moveFromStorage(IStorage $sourceStorage, $sourceInternalPath, $t
 	}
 
 	public function writeStream(string $path, $stream, int $size = null): int {
-		return (int)file_put_contents($this->getSourcePath($path), $stream);
+		$result = file_put_contents($this->getSourcePath($path), $stream);
+		if ($result === false) {
+			throw new GenericFileException("Failed write steam to $path");
```
Comment (kesselb) [inline lib/private/Files/Storage/Local.php:559]: Thanks :+1:

File: lib/private/Files/Storage/Common.php [hunk 3]
```diff
@@ -620,18 +621,21 @@ public function copyFromStorage(IStorage $sourceStorage, $sourceInternalPath, $t
...
 		} else {
 			$source = $sourceStorage->fopen($sourceInternalPath, 'r');
-			// TODO: call fopen in a way that we execute again all storage wrappers
-			// to avoid that we bypass storage wrappers which perform important actions
-			// for this operation. Same is true for all other operations which
-			// are not the same as the original one.Once this is fixed we also
-			// need to adjust the encryption wrapper.
-			$target = $this->fopen($targetInternalPath, 'w');
-			[, $result] = \OC_Helper::streamCopy($source, $target);
+			if ($source) {
```
Comment (kesselb) [inline lib/private/Files/Storage/Common.php:624]: Code would be easier to read if $result is defined with false at the beginning and only set to true on success.

Comment (icewind1991) [reply lib/private/Files/Storage/Common.php:624]: done

Comment (kesselb) [reply lib/private/Files/Storage/Common.php:625]: Thanks @icewind1991 :+1:

**Review by rullzer — APPROVED** [review]
Looks good!

File: lib/private/Files/Storage/Common.php [hunk 4]
```diff
@@ -858,10 +860,13 @@ public function needsPartFile() {
 	public function writeStream(string $path, $stream, int $size = null): int {
 		$target = $this->fopen($path, 'w');
 		if (!$target) {
-			return 0;
+			throw new GenericFileException("Failed to open $path for writing");
 		}
 		try {
 			[$count, $result] = \OC_Helper::streamCopy($stream, $target);
+			if (!$result) {
+				throw new GenericFileException("Failed to copy stream");
```
Comment (MorrisJobke) [inline lib/private/Files/Storage/Common.php:868]:
```suggestion
				throw new GenericFileException("Failed to copy stream to $path");
```

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍 I refined one exception message to also include the path so that looking at the logs makes more sense later on.

**Comment by MorrisJobke** [comment]
PHPUnit fails

**Comment by icewind1991** [comment]
ci fixed

**Comment by MorrisJobke** [comment]
So this breaks deleting files in an object store setup :/

**Comment by MorrisJobke** [comment]
```json
{
    "reqId": "8eOJATfs8cPyjKQ5j2DI",
    "level": 4,
    "time": "2020-07-30T12:30:26+00:00",
    "remoteAddr": "127.0.0.1",
    "user": "admin",
    "app": "webdav",
    "method": "DELETE",
    "url": "/remote.php/dav/files/admin/Bildschirmfoto%202020-07-29%20um%2017.50.40.png",
    "message": {
        "Exception": "TypeError",
        "Message": "Argument 1 passed to OC\\Files\\Cache\\CacheQueryBuilder::whereFileId() must be of the type int, null given, called in /Users/morris/Projects/nextcloud/server/lib/private/Files/Cache/Cache.php on line 661",
        "Code": 0,
        "Trace": [
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Cache/Cache.php",
                "line": 661,
                "function": "whereFileId",
                "class": "OC\\Files\\Cache\\CacheQueryBuilder",
                "type": "->",
                "args": [
                    null
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Cache/Cache.php",
                "line": 585,
                "function": "moveFromCache",
                "class": "OC\\Files\\Cache\\Cache",
                "type": "->",
                "args": [
                    {
                        "__class__": "OC\\Files\\Cache\\Cache"
                    },
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/ObjectStore/ObjectStoreStorage.php",
                "line": 347,
                "function": "move",
                "class": "OC\\Files\\Cache\\Cache",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Storage/Wrapper/Wrapper.php",
                "line": 278,
                "function": "rename",
                "class": "OC\\Files\\ObjectStore\\ObjectStoreStorage",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Storage/Wrapper/Availability.php",
                "line": 295,
                "function": "rename",
                "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Storage/Wrapper/Wrapper.php",
                "line": 278,
                "function": "rename",
                "class": "OC\\Files\\Storage\\Wrapper\\Availability",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Storage/Wrapper/Wrapper.php",
                "line": 572,
                "function": "rename",
                "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/files_trashbin/lib/Trashbin.php",
                "line": 288,
                "function": "moveFromStorage",
                "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                "type": "->",
                "args": [
                    {
                        "cache": {
                            "__class__": "OC\\Files\\Cache\\Cache"
                        },
                        "scanner": null,
                        "watcher": null,
                        "propagator": {
                            "__class__": "OC\\Files\\Cache\\Propagator"
                        },
                        "updater": {
                            "__class__": "OC\\Files\\Cache\\Updater"
                        },
                        "__class__": "OCA\\Files_Trashbin\\Storage"
                    },
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "files_trashbin/files/Bildschirmfoto 2020-07-29 um 17.50.40.png.d1596112226"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/files_trashbin/lib/Trash/LegacyTrashBackend.php",
                "line": 108,
                "function": "move2trash",
                "class": "OCA\\Files_Trashbin\\Trashbin",
                "type": "::",
                "args": [
                    "Bildschirmfoto 2020-07-29 um 17.50.40.png"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/files_trashbin/lib/Trash/TrashManager.php",
                "line": 103,
                "function": "moveToTrash",
                "class": "OCA\\Files_Trashbin\\Trash\\LegacyTrashBackend",
                "type": "->",
                "args": [
                    {
                        "cache": {
                            "__class__": "OC\\Files\\Cache\\Cache"
                        },
                        "scanner": null,
                        "watcher": null,
                        "propagator": {
                            "__class__": "OC\\Files\\Cache\\Propagator"
                        },
                        "updater": {
                            "__class__": "OC\\Files\\Cache\\Updater"
                        },
                        "__class__": "OCA\\Files_Trashbin\\Storage"
                    },
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/files_trashbin/lib/Storage.php",
                "line": 192,
                "function": "moveToTrash",
                "class": "OCA\\Files_Trashbin\\Trash\\TrashManager",
                "type": "->",
                "args": [
                    {
                        "cache": {
                            "__class__": "OC\\Files\\Cache\\Cache"
                        },
                        "scanner": null,
                        "watcher": null,
                        "propagator": {
                            "__class__": "OC\\Files\\Cache\\Propagator"
                        },
                        "updater": {
                            "__class__": "OC\\Files\\Cache\\Updater"
                        },
                        "__class__": "OCA\\Files_Trashbin\\Storage"
                    },
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/files_trashbin/lib/Storage.php",
                "line": 99,
                "function": "doDelete",
                "class": "OCA\\Files_Trashbin\\Storage",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    "unlink"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/View.php",
                "line": 1161,
                "function": "unlink",
                "class": "OCA\\Files_Trashbin\\Storage",
                "type": "->",
                "args": [
                    "files/Bildschirmfoto 2020-07-29 um 17.50.40.png"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/lib/private/Files/View.php",
                "line": 718,
                "function": "basicOperation",
                "class": "OC\\Files\\View",
                "type": "->",
                "args": [
                    "unlink",
                    "/Bildschirmfoto 2020-07-29 um 17.50.40.png",
                    [
                        "delete"
                    ]
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/dav/lib/Connector/Sabre/File.php",
                "line": 455,
                "function": "unlink",
                "class": "OC\\Files\\View",
                "type": "->",
                "args": [
                    "/Bildschirmfoto 2020-07-29 um 17.50.40.png"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/3rdparty/sabre/dav/lib/DAV/Tree.php",
                "line": 183,
                "function": "delete",
                "class": "OCA\\DAV\\Connector\\Sabre\\File",
                "type": "->",
                "args": []
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/3rdparty/sabre/dav/lib/DAV/CorePlugin.php",
                "line": 295,
                "function": "delete",
                "class": "Sabre\\DAV\\Tree",
                "type": "->",
                "args": [
                    "files/admin/Bildschirmfoto 2020-07-29 um 17.50.40.png"
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/3rdparty/sabre/event/lib/WildcardEmitterTrait.php",
                "line": 89,
                "function": "httpDelete",
                "class": "Sabre\\DAV\\CorePlugin",
                "type": "->",
                "args": [
                    {
                        "__class__": "Sabre\\HTTP\\Request"
                    },
                    {
                        "__class__": "Sabre\\HTTP\\Response"
                    }
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/3rdparty/sabre/dav/lib/DAV/Server.php",
                "line": 474,
                "function": "emit",
                "class": "Sabre\\DAV\\Server",
                "type": "->",
                "args": [
                    "method:DELETE",
                    [
                        {
                            "__class__": "Sabre\\HTTP\\Request"
                        },
                        {
                            "__class__": "Sabre\\HTTP\\Response"
                        }
                    ]
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/3rdparty/sabre/dav/lib/DAV/Server.php",
                "line": 251,
                "function": "invokeMethod",
                "class": "Sabre\\DAV\\Server",
                "type": "->",
                "args": [
                    {
                        "__class__": "Sabre\\HTTP\\Request"
                    },
                    {
                        "__class__": "Sabre\\HTTP\\Response"
                    }
                ]
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/3rdparty/sabre/dav/lib/DAV/Server.php",
                "line": 319,
                "function": "start",
                "class": "Sabre\\DAV\\Server",
                "type": "->",
                "args": []
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/dav/lib/Server.php",
                "line": 325,
                "function": "exec",
                "class": "Sabre\\DAV\\Server",
                "type": "->",
                "args": []
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/apps/dav/appinfo/v2/remote.php",
                "line": 35,
                "function": "exec",
                "class": "OCA\\DAV\\Server",
                "type": "->",
                "args": []
            },
            {
                "file": "/Users/morris/Projects/nextcloud/server/remote.php",
                "line": 167,
                "args": [
                    "/Users/morris/Projects/nextcloud/server/apps/dav/appinfo/v2/remote.php"
                ],
                "function": "require_once"
            }
        ],
        "File": "/Users/morris/Projects/nextcloud/server/lib/private/Files/Cache/CacheQueryBuilder.php",
        "Line": 66,
        "CustomMessage": "--"
    },
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Safari/605.1.15",
    "version": "20.0.0.0"
}
```

The `get()` method of the cache returns `false`:

https://github.com/nextcloud/server/blob/cb057829f72c70e819f456edfadbb29d72dba832/lib/private/Files/Cache/Cache.php#L614

Bildschirmfoto 2020-07-30 um 14 41 34

**Comment by acsfer** [comment]
Fix https://github.com/nextcloud/server/issues/19171

**Pull request merged**